File: src/main/java/spring/webmvc/presentation/exception/handler/AccessDeniedExceptionHandler.java
package spring.webmvc.presentation.exception.handler;

import java.io.IOException;

import org.springframework.http.HttpStatus;
import org.springframework.http.ProblemDetail;
import org.springframework.security.access.AccessDeniedException;
import org.springframework.security.web.access.AccessDeniedHandler;
import org.springframework.stereotype.Component;

import jakarta.servlet.http.HttpServletRequest;
import jakarta.servlet.http.HttpServletResponse;
import lombok.RequiredArgsConstructor;
import spring.webmvc.infrastructure.util.ProblemDetailUtil;
import spring.webmvc.infrastructure.util.ResponseWriter;

@Component
@RequiredArgsConstructor
public class AccessDeniedExceptionHandler implements AccessDeniedHandler {

	private final ProblemDetailUtil problemDetailUtil;
	private final ResponseWriter responseWriter;

	@Override
	public void handle(
		HttpServletRequest request,
		HttpServletResponse response,
		AccessDeniedException exception
	) throws IOException {
		ProblemDetail problemDetail = ProblemDetail.forStatusAndDetail(HttpStatus.FORBIDDEN, exception.getMessage());
		problemDetail.setType(problemDetailUtil.createType(HttpStatus.FORBIDDEN));

		responseWriter.writeResponse(problemDetail);
	}
}

File: spring-web-kotlin/src/main/kotlin/spring/web/kotlin/presentation/exception/handler/AccessDeniedExceptionHandler.kt
package spring.web.kotlin.presentation.exception.handler

import jakarta.servlet.http.HttpServletRequest
import jakarta.servlet.http.HttpServletResponse
import org.springframework.http.HttpStatus
import org.springframework.http.ProblemDetail
import org.springframework.security.access.AccessDeniedException
import org.springframework.security.web.access.AccessDeniedHandler
import org.springframework.stereotype.Component
import spring.web.kotlin.infrastructure.util.ResponseWriter

@Component
class AccessDeniedExceptionHandler(
    private val responseWriter: ResponseWriter
) : AccessDeniedHandler {
    override fun handle(
        request: HttpServletRequest?,
        response: HttpServletResponse?,
        accessDeniedException: AccessDeniedException?
    ) {
        responseWriter.writeResponse(
            ProblemDetail.forStatusAndDetail(
                HttpStatus.FORBIDDEN,
                accessDeniedException?.message
            )
        )
    }
}